arch/risc-v/src/qemu-rv/qemu_rv_mm_init.c

```diff
@@ -252,9 +261,23 @@ void qemu_rv_kernel_mappings(void)
 
 void qemu_rv_mm_init(void)
 {
-  /* Setup the kernel mappings */
+#ifdef CONFIG_SMP
+  irqstate_t flags;
+  flags = spin_lock_irqsave(&g_mm_init_spin);
```

Comment:
do we really need spinlock here? before core0 fully initialize other core should be in power off state or wait the signal from cpu0.

Reply:
@xiaoxiang781216 
Yes, we need the code to run BUILD_KERNEL in SMP, because all cores start at the same time.
In ARMv-7A, @GUIDINGLI recently modified the startup code which uses ARM_SEV to send a signal to all cores.
However, RISC-V does not have such instruction and technically we can send signals by IPI but I think this should be implemented in another PR later.

Reply:
In this case, the second core has to wait until the main cpu finish the basic initialization. The spin lock hold here isn't enough.

Reply:
@xiaoxiang781216 
I have other changes of qemu_rv_start.c for SMP which I was going to commit later, but I think I should add to this PR as well. And regarding initialization of `g_mm_init_spin`, it relies on BSS clearing but I can add the default value explicitly.

Reply:
@xiaoxiang781216 
I've just added one more commit to this PR.

Reply:
qemu_rv_mm_init should be called only on the primary core, so it's better to check hart id equals zero like up_mtimer_initialize.

Reply:
@xiaoxiang781216 
Actually all CPUs need to call `qemu_rv_mm_init` to set the satp register for MMU on each CPU.

Reply:
Yes, satp need done by all cpu, but the construction of page table should be done by primary core only. I mean the code can check hartid to decide whether to skip the shared part. But for the synchronization, we can't depend on the normal variable or spinlock because they mayn't fully initialize if the secondary core run too fast.